Accessibility updates #264 - Add aria attributes to better support screen readers and keyboard navigation

cooperfellows opened the pull request:
<!---
Good pull requests — patches, improvements, new features — are a fantastic help.  They should remain focused in scope and avoid containing unrelated commits.

Please review the Pull Requests section of our Contributing Guidelines before submitting your work: https://github.com/harvesthq/chosen/blob/master/contributing.md#pull-requests
-->
### Summary

This adds a good chunk of ARIA attributes and other related attributes to the chosen elements to move towards making Chosen more accessible. 

**ARIA Labels and Other Helpful Attributes**
- Adds the following aria elements to the search input box
  - [role="combobox"](https://www.w3.org/TR/wai-aria/roles#combobox)
  - [aria-haspopup](https://www.w3.org/TR/wai-aria/states_and_properties#aria-haspopup) (implicit when using the combobox role)
  - [aria-expanded](https://www.w3.org/TR/wai-aria/states_and_properties#aria-expanded)
  - [aria-activedescendant="id_of_highlighted_option"](https://www.w3.org/TR/wai-aria/states_and_properties#aria-activedescendant)
  - [aria-owns="id_of_list_of_options"](https://www.w3.org/TR/wai-aria/states_and_properties#aria-owns)
  - [aria-autocomplete="list"](https://www.w3.org/TR/wai-aria/states_and_properties#aria-autocomplete)
  - [aria-labeledby="id_of_field_label"](https://www.w3.org/TR/wai-aria/states_and_properties#aria-labelledby)
- Adds the following attributes to the list of options
  - id
  - [role="listbox"](https://www.w3.org/TR/wai-aria/roles#listbox)
- Adds the following attributes to each individual option in the list of options
  - id
  - [role="option"](https://www.w3.org/TR/wai-aria/roles#option)
  - [aria-busy](https://www.w3.org/TR/wai-aria/states_and_properties#aria-busy)

**Managing State**
- Managing the aria-expanded attribute
  - In order to indicate when the search results should be relevant to assitive technology, we need to toggle the aria-expanded attribute as fields are activated/deactivated.
  - The easiest way to do this (AFAIK) is to adjust the attribute during the `close_field` and `activate_field` methods.
  - A simple switch from true to false or false to true in each of these methods is sufficient enough to keep the state updated

A complete description of the changes made, with the reasoning behind each attribute is available here: 
- https://github.com/harvesthq/chosen/issues/264#issuecomment-215124540
### References
- [Accessibility issues with chosen #264](https://github.com/harvesthq/chosen/issues/264)

cooperfellows committed:
Add initial round of aria elements to jQuery

Changes inspired by
https://github.com/kirstenmalin/chosen/commit/4d0da63e3839df229c3095e692395acbd39d811a

Use the label element, aria-label, or aria-labelledby attributes from
the select box for the text input field.
Add attributes and ids such that a screen reader can know which element
in the search results is highlighted so that it can be read:
Attributes on the text input:
aria-haspopup="true"
aria-expanded="true" (needs to be changed dynamically as the field is
activated/deactivated)
role="combobox"
aria-autocomplete="list"
aria-owns="id_of_list_of_options"
aria-activedescendant="id_of_highlighted_option" (change this value
programatically when the highlighted option changes)
aria-labeledby="id_of_field_label"
Attributes on the list of options:
id for use in the aria-owns attribute on the input
Attributes on each option:
role="option"
id for use in the aria-activedescendant attribute on the input

cooperfellows committed:
Add ARIA elements in prototype code

Changes inspired by
https://github.com/kirstenmalin/chosen/commit/4d0da63e3839df229c3095e692395acbd39d811a

Use the label element, aria-label, or aria-labelledby attributes from
the select box for the text input field.
Add attributes and ids such that a screen reader can know which element
in the search results is highlighted so that it can be read:
Attributes on the text input:
aria-haspopup="true"
aria-expanded="false" (needs to be changed dynamically as the field is
activated/deactivated)
role="combobox"
aria-autocomplete="list"
aria-owns="id_of_list_of_options"
aria-activedescendant="id_of_highlighted_option" (change this value
programatically when the highlighted option changes)
aria-labeledby="id_of_field_label"
Attributes on the list of options:
id for use in the aria-owns attribute on the input
Attributes on each option:
role="option"
id for use in the aria-activedescendant attribute on the input

cooperfellows committed:
Add new trigger buildresultslist

This is simply a pass through to the winnow_results function. There is
an A11Y reason behind this, when a page loads, and chosen is activated,
the results list is empty, so the aria labels are pointing to an empty
list, which is flagged as an error in many web scanners. As soon as the
results list is built (the first time the field is activated) the
scanners stop complaining.

The reason I chose to add it as a trigger, is for dynamic select lists.
Normally, chosen:updated would be used when options are inserted, the
problem is that you can't use chosen:updated until the results list has
been built once, because it doesn't actually generate the results list.

I have decided NOT to force a build on initialization to avoid adding
overhead for those who don't need the A11Y work. Instead, the trigger
needs to be called in your script wherever Chosen is initialized.

If there is a better way to build the options list than this, please let
me know, I am admittedly new to the Chosen source code.

cooperfellows committed:
Update aria-expanded during activate/close of field

cooperfellows committed:
Add aria-busy to search list

Default to busy since there are no results until the field is activated.
Upon activation of the field, set aria-busy to false.

cooperfellows committed:
Add aria-busy to prototype

Also add missing role="listbox" to chosen-results element

cooperfellows committed:
Remove buildresults trigger

Since we are using aria-busy, there is no reason to build out the
results early like before.

cooperfellows committed:
Update aria-busy selector

Was not hitting the right element in previous commit.

tjschuck commented:
Thanks @cooperfellows!

Just linking a couple more related PRs: #2013 and #2047

cooperfellows commented:
@tjschuck Happy to help, but I think there is a bug in the chosen.prototype.coffee. The build process flagged in in the detailed output, but claims it passed anyway. Let me take a look at it and get back to you.

stof commented:
``.attr`` is a jquery method. Using it in the prototype version is wrong.

stof commented:
btw, you used the proper Prototype methods in other places dealing with attributes, so this looks weird to me.

cooperfellows commented:
Yea, I'm not a Prototype guy, I am seeing some other issues with the attribute values. It looks like I should be passing string "true" and string "false" not the boolean. At least according to:
 - https://github.com/sstephenson/prototype/blob/5fddd3e/src/prototype/dom/dom.js#L2385

Working on an update now.

cooperfellows committed:
Update prototype with correct attribute syntax

Fix incorrect syntax in prototype.coffee file when writing attribute
values.  (.writeAtribute, not .attr)

According to prototype source
(https://github.com/sstephenson/prototype/blob/5fddd3e/src/prototype/dom/dom.js#L2385)
passing boolean false will remove the attribute, and passing boolean
true will set the value of the attribute to the attribute name, not the
value. So pass string "true" and string "false" instead.

cooperfellows committed:
Update missed read attribute value call

Needed to fix prototype readAttribute call. Too used to jQuery...

cooperfellows commented:
I'm sorry this got so ugly. I was trying to finish this up before I take off for a week so I've been rushing things a bit more than I would like to. I missed one other .attr call which I've now corrected. 

Let me know what else you want from my end, happy to help when I can. Thanks for the response and the pointers.

cooperfellows committed:
Ensure controls.labels support exists

Not all browsers are implementing this yet

cooperfellows committed:
Add IE8 Compatible version of hasOwnProperty

miconley commented:
@pfiller Have you and your team taken a look?  My team prefers to use this library but cannot until it's A11y-ready.  Thanks!

arvisto commented:
Hi Cooperfellows, could you give an update on your plans for the future of this pull request?

cooperfellows commented:
Hi @arvisto I haven't been keeping this up to date with any other changes in Chosen. I have been using a version with all the changes I made in this PR in several of our production applications without any known issues.

To be clear though, not hearing about any issues does not mean there are none. Unfortunately I am too busy to make any updates to this for at least the next 6 months (some major projects at work). So for now this will most likely remain stagnant.

arvisto commented:
@cooperfellows Thank you for the update. Hopefully these changes get merged in at some point, chosen would benefit greatly from adhering to accessibility standards, as most Canadian companies that have anything to do with the government have to adhere by law. But I think I'm preaching to the choir.

ckpicker commented:
@cooperfellows Nicely done!  Any idea when this is going to be merged?  I would love to start using this asap.
